### PR TITLE
[Processing] third group of vector geometry algotihms

### LINKED ISSUE #1305: [FEATURE][processing] New algorithm for single sided buffers
Original commit: https://github.com/qgis/QGIS/commit/989d9864c5457e96f861b66695ee2d6124899ae8 by nyalldawson

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1374: [FEATURE][processing] Add choice of simplification method to simplify
Original commit: https://github.com/qgis/QGIS/commit/56b77db88c4b617f254a7fa1309f86ee7d3cedfa by nyalldawson

This change allows users to choose which method to use when running
the simplify geometries algorithm, with choices of the existing
distance based (Douglas Peucker) algorithm, area based (Visvalingam)
algorithm and snap-to-grid.

Visvaligam in particular usually results in more cartographically
pleasing simplification over the standard distance based methods.

### LINKED ISSUE #1376: [FEATURE] Extra control when smoothing geometries
Original commit: https://github.com/qgis/QGIS/commit/4d60d0cdb54cc75f15c4e841aa87a3fd6994c918 by nyalldawson

Add option to QgsGeometry::smooth to not smooth segments shorter than a certain threshold or sharp corners with an angle exceeding a threshold

Expose the angle threshold to processing smooth algorithm

Also:
- optimise QgsGeometry::smooth for new geometry classes
- Fix smooth does not work with geometries containing Z/M

### LINKED ISSUE #1505: [FEATURE][processing] Snap geometries to layer algorithm
Original commit: https://github.com/qgis/QGIS/commit/c3a978b9da33f69c6c5440e1166c78c89946dde4 by nyalldawson

Port the Geometry Snapper plugin across to the analysis lib, and
expose to python bindings

Add a new algorithm which performs the snapping to layers

### LINKED ISSUE #1506: [FEATURE][processing] Snap geometries algorithm allows snapping
to other layer types, supports point/line layers
Original commit: https://github.com/qgis/QGIS/commit/dae0a017617d3082b6e89f8a3e75c2c9bcd0ebf5 by nyalldawson

Fix #14791, #15313

### LINKED ISSUE #1953: [FEATURE] Subdivide algorithm for QgsGeometry
Original commit: https://github.com/qgis/QGIS/commit/e74395d95bd326788ce55f2a7caa31435fed9ea9 by nyalldawson

Subdivides the geometry. The returned geometry will be a collection
containing subdivided parts from the original geometry, where no
part has more then the specified maximum number of nodes.

### LINKED ISSUE #1954: [FEATURE][processing] New algorithm for subdividing geometries
Original commit: https://github.com/qgis/QGIS/commit/4dcf8d8d10249971f5ec1358a172078e11e6bad8 by nyalldawson

The returned geometry will be a collection containing subdivided parts
from the original geometry, where no part has more then the specified
maximum number of nodes.

This is useful for dividing a complex geometry into less complex parts,
which are better able to be spatially indexed and faster to perform
further operations such as intersects on. The returned geometry parts may
not be valid and may contain self-intersections.

### LINKED ISSUE #1992: [FEATURE] New algorithms to add Z/M values to existing geometries
Original commit: https://github.com/qgis/QGIS/commit/340cf93f93fb28e1aa9e23b2d80a7c38e6a89d6c by nyalldawson

Allows upgrading geometries to include these dimensions, or
overwriting any existing Z/M values with a new value.

Intended mostly as a test run for QgsProcessingFeatureBasedAlgorithm

### LINKED ISSUE #2079: [FEATURE] Native 'Promote to Multipart' algorithm
Original commit: https://github.com/qgis/QGIS/commit/3484eb019c2b8eb682e5a41ff9e6e4b14b9e97d4 by nyalldawson

This algorithm is basically the equivalent of the ST_Multi(...)
command - it forces a feature's geometry to become multipart,
regardless of the input geometry type.

If input geometries are singlepart, they will output as
multipart with just 1 part. If they are already multipart,
they will be output unchanged.

### LINKED ISSUE #2081: [FEATURE] Remove Singleparts to Multiparts algorithm
Original commit: https://github.com/qgis/QGIS/commit/a55fbd8ef349a52c166b69b70ce6cfccad8c42aa by nyalldawson

This algorithm is no longer required - it's been replaced by
the 'Promote to multipart' and 'Collect geometries" algorithms.

Tagged as feature to remember to include in release notes

### PR DESCRIPTION
**Pull request by ghtmtt**
This is the third (and last) of 3 PR for the Processing Vector Geometry section. Supersedes #2240 

Fix #2079 fix #2081 fix #1992 fix #1953 fix #1954 fix #1505 fix #1506 fix #1376 fix #1374 fix #1305

reviewers are super welcome!

### CODE REVIEW

**Comment by ghtmtt** [comment]
Travis fails because in this part a picture come from #2255 
Should I add the picture also in this PR?

BTW: I build local html and I did get any error

**Comment by DelazJ** [comment]
If you did a [quick build](https://github.com/qgis/QGIS-Documentation#speed-up-the-documentation-build) then the images could have been in "cache" from the previous build and it processed fine.
> Should I add the picture also in this PR?

Yes. Better have pictures with the algs they refer to in the same PR. and remove them from the one they do not belong to (in case the picture has to evolve, it avoids duplicating changes and/or overwriting inadvertently the one that has been approved when merging)

**Comment by ghtmtt** [comment]
pictures are there. The failing is due to missing anchors because of the splitting of the PR. After merging there shouldn't be other problems.

**Comment by ghtmtt** [comment]
Fixes #2079 #2081 #1992 #1953 #1954 #1505 #1506 #1376 #1374 #1305

File: source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst [hunk 1]
```diff
@@ -907,219 +907,745 @@ Console usage
...
 ``Segments`` [number]
-  <put parameter description here>
+  Controls the number of line segments to use to approximate a quarter circle when
+  creating rounded offsets
 
   Default: *5*
 
 ``Dissolve result`` [boolean]
-  <put parameter description here>
+  Choose to dissolve the final buffer
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst]: Choose to dissolve the final buffer, resulting in a single feature covering all input features.

File: source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst [hunk 2]
```diff
@@ -907,219 +907,745 @@ Console usage
...
 
 ``Input layer`` [vector: point]
-  <put parameter description here>
+  Input point vector layer
 
 ``Buffer region`` [number]
-  <put parameter description here>
-
-  Default: *0.0*
+  Allow to extend the area of the Voronoi polygons
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst]: area of the Voronoi polygons or of the input layer (extending its extent)?

File: source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst [hunk 3]
```diff
@@ -907,219 +907,745 @@ Console usage
...
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst]: Vorony Buffer region has also a default value 0.0

**Pull request closed without merging**